diesel/src/sqlite/connection/raw.rs

```diff
@@ -194,6 +243,115 @@ extern "C" fn run_custom_function<F>(
+// See: https://github.com/rust-lang/rfcs/blob/master/text/2195-really-tagged-unions.md#opaque-tags
+#[repr(u8)]
+enum OptionalAggregator<A> {
+    // Discriminant is 0
+    #[allow(dead_code)]
+    None,
+    Some(A),
+}
+
+#[allow(warnings)]
+extern "C" fn run_aggregator_step_function<ArgsSqlType, RetSqlType, Args, Ret, A>(
+    ctx: *mut ffi::sqlite3_context,
+    num_args: libc::c_int,
+    value_ptr: *mut *mut ffi::sqlite3_value,
+) where
+    A: SqliteAggregateFunction<Args, Output = Ret> + 'static + Send,
+    Args: Queryable<ArgsSqlType, Sqlite>,
+    Ret: ToSql<RetSqlType, Sqlite>,
+    Sqlite: HasSqlType<RetSqlType>,
+{
+    unsafe {
+        let aggregate_context = ffi::sqlite3_aggregate_context(
+            ctx,
+            std::mem::size_of::<OptionalAggregator<A>>() as i32,
+        );
+        let mut aggregate_context = NonNull::new(aggregate_context as *mut OptionalAggregator<A>);
+        let aggregator = match aggregate_context.map(|a| &mut *a.as_ptr()) {
+            Some(&mut OptionalAggregator::Some(ref mut agg)) => agg,
+            Some(mut a_ptr @ &mut OptionalAggregator::None) => {
+                ptr::write_unaligned(a_ptr as *mut _, OptionalAggregator::Some(A::default()));
+                if let &mut OptionalAggregator::Some(ref mut agg) = a_ptr {
+                    agg
+                } else {
+                    unreachable!(
+                        "We've written the aggregator above to that location, it must be there"
+                    )
+                }
+            }
+            None => {
+                null_aggregate_context_error(ctx);
+                return;
+            }
+        };
+
+        let mut f = |args: &[*mut ffi::sqlite3_value]| -> Result<(), Error> {
+            let args = build_sql_function_args::<ArgsSqlType, Args>(args)?;
+
+            Ok(aggregator.step(args))
+        };
+
+        let args = slice::from_raw_parts(value_ptr, num_args as _);
+        match f(args) {
+            Err(e) => {
+                let msg = e.to_string();
+                ffi::sqlite3_result_error(ctx, msg.as_ptr() as *const _, msg.len() as _);
+            }
+            _ => (),
+        };
+    }
+}
+
+extern "C" fn run_aggregator_final_function<ArgsSqlType, RetSqlType, Args, Ret, A>(
+    ctx: *mut ffi::sqlite3_context,
+) where
+    A: SqliteAggregateFunction<Args, Output = Ret> + 'static + Send,
+    Args: Queryable<ArgsSqlType, Sqlite>,
+    Ret: ToSql<RetSqlType, Sqlite>,
+    Sqlite: HasSqlType<RetSqlType>,
+{
+    unsafe {
+        let aggregate_context = ffi::sqlite3_aggregate_context(ctx, 0);
+        let mut aggregate_context = NonNull::new(aggregate_context as *mut OptionalAggregator<A>);
+        let aggregator = match aggregate_context {
+            Some(ref mut a) => match std::mem::replace(a.as_mut(), OptionalAggregator::None) {
+                OptionalAggregator::Some(agg) => agg,
+                OptionalAggregator::None => {
```

Comment:
If I think about this case more in detail, this happens if you call your aggregate function on an empty set. So we should probably just return a null value in this cases? 
That implies a few more things:
* We should have a test case covering applying an aggregate function to an empty select.
* If we decide to return null in such cases, that means we need to enforce that we only allow to register custom aggregate functions for function definitions where the return type is `Nullable<…>`.
* We should probably indicate this also in the definition of `SqliteAggregateFunction` by making `finalize` explicitly return a `Option<Self::Output>`

(Sorry for not noticing that any earlier :see_no_evil:)

Reply:
Sounds good! I should've thought about that as well early on